sycl/source/detail/kernel_bundle_impl.hpp

```diff
@@ -796,21 +814,21 @@ class kernel_bundle_impl {
 
   bool is_specialization_constant_set(const char *SpecName) const noexcept {
     bool SetInDevImg =
-        std::any_of(MDeviceImages.begin(), MDeviceImages.end(),
+        std::any_of(MUniqueDeviceImages.begin(), MUniqueDeviceImages.end(),
                     [SpecName](const device_image_plain &DeviceImage) {
                       return getSyclObjImpl(DeviceImage)
                           ->is_specialization_constant_set(SpecName);
                     });
     return SetInDevImg || MSpecConstValues.count(std::string{SpecName}) != 0;
   }
 
-  const device_image_plain *begin() const { return MDeviceImages.data(); }
+  const device_image_plain *begin() const { return MUniqueDeviceImages.data(); }
```

Comment:
So `has_kernel()` now uses just plain `begin()` and `end()` .  But `is_specialization_constant_set` uses the more explicit `MUniqueDeviceImages.begin()/end()`.  It seems that they should be consistent.  But which way is better?

I kind of feel that we should either remove the `begin()` and `end()` functions from `kernel_bundle_impl` and always just use the explicit call, so it's clear when we are enumerating the `MUniqueDeviceImages` versus the `MDeviceImages`.

But, on the other hand, it's clear that `MUniqueDeviceImages` is the "go to" default, in which case, we should keep these top level `begin() / end()` functions and instead have `is_specialization_constant_set` (and similar functions) use the top level `begin()` and `end()` calls, rather than explicit enumerations of `MUniqueDeviceImages` .  

What do you think?